long-ride-alerts/src/test/java/org/apache/flink/training/exercises/longrides/LongRidesTest.java

```diff
@@ -18,89 +18,150 @@
-        TaxiRide rideEnded3HoursLater = endRide(rideStarted, BEGINNING.plusSeconds(180 * 60));
 
-        TestRideSource source =
-                new TestRideSource(rideStarted, mark2HoursLater, rideEnded3HoursLater);
-        assertEquals(Collections.singletonList(rideStarted), results(source));
+        ParallelTestSource<TaxiRide> source =
+                new ParallelTestSource<>(new PartitionByRideId(), rideStarted);
+        TestSink<Long> sink = new TestSink<Long>();
+
+        longRidesPipeline().execute(source, sink);
+        assertThat(sink.results()).containsExactly(rideStarted.rideId);
     }
 
     @Test
-    public void startIsDelayedMoreThanTwoHours() throws Exception {
-        TaxiRide rideStarted = startRide(1, BEGINNING);
-        TaxiRide rideEndedAfter1Hour = endRide(rideStarted, BEGINNING.plusSeconds(60 * 60));
-        Long mark2HoursAfterEnd = BEGINNING.plusSeconds(180 * 60).toEpochMilli();
+    public void endComesAfter3Hours() throws Exception {
+        TaxiRide startOfLongRide = startRide(1, BEGINNING);
+        TaxiRide longRideEndedAfter3Hours = endRide(startOfLongRide, THREE_HOURS_LATER);
 
-        TestRideSource source =
-                new TestRideSource(rideEndedAfter1Hour, mark2HoursAfterEnd, rideStarted);
-        assert (results(source).isEmpty());
+        ParallelTestSource<TaxiRide> source =
+                new ParallelTestSource<>(
+                        new PartitionByRideId(), startOfLongRide, longRideEndedAfter3Hours);
+        TestSink<Long> sink = new TestSink<Long>();
+
+        longRidesPipeline().execute(source, sink);
+        assertThat(sink.results()).containsExactly(startOfLongRide.rideId);
+    }
+
+    @Test
+    public void multipleRides() throws Exception {
+        TaxiRide startOfOneRide = startRide(1, BEGINNING);
+        TaxiRide otherRide = startRide(2, ONE_MINUTE_LATER);
+        TaxiRide oneRideEnded = endRide(startOfOneRide, THREE_HOURS_LATER);
+        TaxiRide otherRideEnded = endRide(otherRide, THREE_HOURS_LATER);
+
+        ParallelTestSource<TaxiRide> source =
+                new ParallelTestSource<>(
+                        new PartitionByRideId(),
+                        startOfOneRide,
+                        otherRide,
+                        oneRideEnded,
+                        otherRideEnded);
+        TestSink<Long> sink = new TestSink<Long>();
+
+        longRidesPipeline().execute(source, sink);
+        assertThat(sink.results())
+                .containsExactlyInAnyOrder(startOfOneRide.rideId, otherRide.rideId);
     }
 
-    private TaxiRide testRide(long rideId, Boolean isStart, Instant startTime, Instant endTime) {
+    // Arranges for all events for a given rideId to be generated by the same source subtask.
+    private static class PartitionByRideId implements TestSourcePartitioner<TaxiRide> {
+        @Override
+        public long partition(TaxiRide ride) {
+            return ride.rideId;
+        }
+    }
```

Comment:
Why is this important? Wouldn't it be better that things are arbitrarily distributed between source "partitions"?

Reply:
You're right. This is a vestige of something that I thought I needed, but then mostly abandoned. I should have removed it, which I've just done.